snuba/utils/streams/kafka.py

```diff
@@ -630,3 +638,71 @@ def close(self, timeout: Optional[float] = None) -> None:
         messages: int = self.__producer.flush(*[timeout] if timeout is not None else [])
         if messages > 0:
             raise TimeoutError(f"{messages} commit log messages pending delivery")
+
+
+class KafkaProducer(Producer[TPayload]):
+    def __init__(
+        self, configuration: Mapping[str, Any], codec: Codec[KafkaPayload, TPayload]
+    ) -> None:
+        self.__configuretion = configuration
+        self.__codec = codec
+
+        self.__producer = ConfluentProducer(configuration)
+        self.__shutdown_requested = Event()
+
+        self.__result = execute(self.__worker)
+
+    def __worker(self) -> None:
+        while not self.__shutdown_requested.is_set():
+            self.__producer.poll(0.1)
+        self.__producer.flush()
+
+    def __delivery_callback(
+        self,
+        future: Future[MessageDetails],
+        error: KafkaError,
+        message: ConfluentMessage,
+    ) -> None:
+        if error is not None:
+            future.set_exception(TransportError(error))
+        else:
+            future.set_result(
+                MessageDetails(
+                    Partition(Topic(message.topic()), message.partition()),
+                    message.offset(),
+                )
+            )
+
+    def produce(
+        self, destination: Union[Topic, Partition], payload: TPayload
+    ) -> Future[MessageDetails]:
+        if self.__shutdown_requested.is_set():
+            raise RuntimeError("producer has been closed")
+
+        if isinstance(destination, Topic):
+            produce = partial(self.__producer.produce, topic=destination.name)
+        elif isinstance(destination, Partition):
+            produce = partial(
+                self.__producer.produce,
+                topic=destination.topic.name,
+                partition=destination.index,
+            )
+        else:
+            raise TypeError("invalid destination type")
+
+        future: Future[MessageDetails] = Future()
+        future.set_running_or_notify_cancel()
+        try:
+            encoded = self.__codec.encode(payload)
```

Comment:
Thinking about this more, I think the encoding step should happen outside of the `try` block so that it throws immediately, instead of being bound to the future. (This came out of some thinking here: https://github.com/getsentry/snuba/pull/669#discussion_r364034169)

Reply:
Ended up just removing this entire block, that way if there are any issues prior to actually putting the bytes on the network (encoding errors, full producer queue, etc.) it raises immediately.